Generalize functions to apply to stacks of images

TAdeJong opened the pull request:
In anticipation of an attempt to get this work into skimage, I wrote an extension to make the functions applicable to stacks of images. 

By adding an axes parameter to all flavors of the per operator.
Note: This still only applies the transform to 2D slices of the nD array
and is no true generalization to N-dimensional datasets (the original
paper does not treat the nD case, although the extension seems
straightforward mathematically.

As a next step I will update the docstring to conform to the NumPy standard, if it is desirable I can add that to this same pull request.

TAdeJong committed:
Generalize functions to apply to stacks of images

By adding a axes parameter to all flavors of the per operator.
Note: This still only applies the transform to 2D slices of the nD array
and is no true generalization to N-dimensional datasets (the original
paper does not treat the nD case, although thge extension seems
straightforward to implement.

TAdeJong committed:
Update docstring to numpy formatting

sbrisard commented:
Hi Tobias,
thank you for your interest in this project. Unfortunately, I do not have time to review your PR at the moment. Forking seems to be a good option. Then, feel free to contribute your fork to `skimage` if you want to go down this route. I think the license of the original project covers this case, and I am happy to send you any paperwork (if required) for inclusion in `skimage`.
What I will kindly ask from you (no obligation) is some credit. Also, do speak around about my blog, if you think it's worth it. It's not very active, but I will certainly write more if I get more readers.
Meanwhile, extension to nD is doable. I've done it in 3D for [this paper](https://hal-enpc.archives-ouvertes.fr/hal-01904481) and am happy to help if necessary (I can even send you the 3D code).
Best regards,
Sébastien

TAdeJong committed:
Fix SyntaxWarning: "is" with a literal in Python3.8+

TAdeJong committed:
Merge pull request #1 from TAdeJong/skimage-prep

Skimage prep

sbrisard commented:
Hi Tobias,
thanks for this update.
I'm hoping to review the whole PR soon. Thanks for your patience!

sbrisard commented on moisan2011.py:
```diff
@@ -137,10 +158,22 @@ def per(u, inverse_dft=True):
     u = np.asarray(u, dtype=np.float64)
 
-    m, n = u.shape
+    if axes == (-2, -1):
+        ut = u
+    else:
```
Is this `if` really necessary? It does not cover all cases where the `moveaxis` is not necessary, and skeeping this `moveaxis` when it is not required probably leads to a negligible speed gain.

Besides, this test is not carried out in the `return` clause.

TAdeJong replied:
agreed and fixed everywhere

sbrisard commented on moisan2011.py:
```diff
@@ -187,41 +243,76 @@ def per2(u, inverse_dft=True):
     """
-    u = np.asarray(u, dtype=np.float64)
-    m, n = u.shape
+    if axes == (-2, -1):
+        ut = u
+    else:
```
I would remove (see above).

sbrisard commented on moisan2011.py:
```diff
@@ -231,92 +322,141 @@ def rper(u, inverse_dft=True):
     """
     u = np.asarray(u, dtype=np.float64)
-    m, n = u.shape
+    if axes == (-2, -1):
+        ut = u
+    else:
```
See above

sbrisard commented on moisan2011.py:
```diff
@@ -231,92 +322,141 @@ def rper(u, inverse_dft=True):
     If inverse_dft is False, then the pair
 
-        (numpy.fft.rfft2(p), numpy.fft.rfft2(s))
+        (numpy.fft.rfft2(p axes=axes), numpy.fft.rfft2(s, axes=axes))
```
Missing comma?

TAdeJong replied:
missing comma indeed. Fixed.

sbrisard commented on moisan2011.py:
```diff
@@ -231,92 +322,141 @@ def rper(u, inverse_dft=True):
-    u = np.asarray(u, dtype=np.float64)
-    m, n = u.shape
+    if axes == (-2, -1):
+        ut = u
```
See above

sbrisard commented:
Hi Tobias,

thanks a lot for your work. I have only minor remarks (see above and feel free to disagree). Onve we agree with these, your PR is ready to be merged.

Looking at your PR, a few remarks came to my mind

- the original code shows some duplications in the implementation of the complex/real operators
- extension to multi-dimensions is indeed easy (I've done it for 3D in a research project), but the pure Fourier implementation will require a little bit of book keeping because of the trick that uses 1D FFTs to precompute the kernel.

What do you think?

TAdeJong committed:
Remove transpose check, add explicit asarray

Per @sbrisard 's comments in
https://github.com/sbrisard/moisan2011/pull/1.
Numpy is smart and fast enough to not need to check for a no-op
transpose
Add explicit asarray with np.float64 in the lesser used functions.

TAdeJong commented:
Hi Sébastien,

I agree with all your comments and fixed them. Unfortunately, I can not spend too much time on this at the moment as I am trying to finish a PhD dissertation. 

- I do agree that there is some duplication between real and complex operators, but I don't think it is a real problem at this point. 
- I personally do not have a use case for nD in my current work, but I think having a scipy implementation of it in the public domain would be cool. scikit-image does encourage / aim to have everything work for nD images, that's why I brought it up. However, I can also imagine building a pull request with the current 2D code and see where we end up from there. Unfortunately, I do not have the spare time to build such a pull request at the moment.
- While updating this branch, I noticed 6 tolerance tests on `per2` and `rper2` are failing by about 6 orders of magnitude for both this branch as well as the current master. Is this a real problem, is it a known problem, or should the tolerances just be adjusted?
- Talking about testing, I could quickly setup a github actions workflow to run `flake8` and the tests automatically for this repository if you like?

sbrisard commented:
Thanks for your changes, Tobias.
I will not go through the submission process to `skimage` myself.
For the time being, I will merge your PR. If you still have time to deal with this task in a few months, feel free to do so.
As for code duplications and going to nd, I was merely mentioning these points.
What you indicated about the tests is rather worrying to me. I usually do not commit tests until they pass. I will enquire.
I'd rather not set github actions up for the time being (not sure I would be able to maintain them). But I should certainly run `black` once.

TAdeJong commented:
I attached the full pytest output, it maybe helpful. I had it reproduce on 2 machines (I was concerned my use of `nomkl` could be the cause, but it wasn't), but in admittedly somewhat arbitrary (python3) environments.

<details>
<summary>Full pytest output</summary>

```
$ pytest .
============================================ test session starts ============================================
platform linux -- Python 3.7.3, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/tobias/code/moisan2011
collected 144 items                                                                                         

test_moisan2011.py .................................................................................. [ 56%]
...F...F...........F...F.......................F...F..........                                        [100%]

================================================= FAILURES ==================================================
____________________________________ test_mean_s_is_zero[per2-u17-False] ____________________________________

per = <function per2 at 0x7fd35cc2d6a8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 78.,  63.,  59.],
       [122., 107.,  87., ...,  76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft',
                             itertools.product(operators(), real_images(),
                                               [True, False]))
    def test_mean_s_is_zero(per, u, inverse_dft):
        _, s = per(u, inverse_dft=inverse_dft)
        if inverse_dft:
            assert np.abs(np.mean(s)) <= 1E-13
        else:
>           assert np.abs(s[0, 0]) <= 1E-15
E           AssertionError: assert 3.841975653468968e-09 <= 1e-15
E            +  where 3.841975653468968e-09 = <ufunc 'absolute'>((-3.725290298461914e-09+9.396750044743385e-10j))
E            +    where <ufunc 'absolute'> = np.abs

test_moisan2011.py:100: AssertionError
____________________________________ test_mean_s_is_zero[per2-u21-False] ____________________________________

per = <function per2 at 0x7fd35cc2d6a8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.],
       [113.,  94., 105., ...,  77.,  81.,  74.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft',
                             itertools.product(operators(), real_images(),
                                               [True, False]))
    def test_mean_s_is_zero(per, u, inverse_dft):
        _, s = per(u, inverse_dft=inverse_dft)
        if inverse_dft:
            assert np.abs(np.mean(s)) <= 1E-13
        else:
>           assert np.abs(s[0, 0]) <= 1E-15
E           AssertionError: assert 3.842588143336527e-09 <= 1e-15
E            +  where 3.842588143336527e-09 = <ufunc 'absolute'>((-3.725290298461914e-09+9.42176114904214e-10j))
E            +    where <ufunc 'absolute'> = np.abs

test_moisan2011.py:100: AssertionError
___________________________________ test_mean_s_is_zero[rper2-u33-False] ____________________________________

per = <function rper2 at 0x7fd35cc2d7b8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 78.,  63.,  59.],
       [122., 107.,  87., ...,  76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft',
                             itertools.product(operators(), real_images(),
                                               [True, False]))
    def test_mean_s_is_zero(per, u, inverse_dft):
        _, s = per(u, inverse_dft=inverse_dft)
        if inverse_dft:
            assert np.abs(np.mean(s)) <= 1E-13
        else:
>           assert np.abs(s[0, 0]) <= 1E-15
E           AssertionError: assert 3.725290298461914e-09 <= 1e-15
E            +  where 3.725290298461914e-09 = <ufunc 'absolute'>((-3.725290298461914e-09+0j))
E            +    where <ufunc 'absolute'> = np.abs

test_moisan2011.py:100: AssertionError
___________________________________ test_mean_s_is_zero[rper2-u37-False] ____________________________________

per = <function rper2 at 0x7fd35cc2d7b8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.],
       [113.,  94., 105., ...,  77.,  81.,  74.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft',
                             itertools.product(operators(), real_images(),
                                               [True, False]))
    def test_mean_s_is_zero(per, u, inverse_dft):
        _, s = per(u, inverse_dft=inverse_dft)
        if inverse_dft:
            assert np.abs(np.mean(s)) <= 1E-13
        else:
>           assert np.abs(s[0, 0]) <= 1E-15
E           AssertionError: assert 3.725290298461914e-09 <= 1e-15
E            +  where 3.725290298461914e-09 = <ufunc 'absolute'>((-3.725290298461914e-09+0j))
E            +    where <ufunc 'absolute'> = np.abs

test_moisan2011.py:100: AssertionError
______________________________________ test_consistency[per2-u9-False] ______________________________________

per = <function per2 at 0x7fd35cc2d6a8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 78.,  63.,  59.],
       [122., 107.,  87., ...,  76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft', consistency_params())
    def test_consistency(per, u, inverse_dft):
        eps = {moisan2011.per: 1E-11,
               moisan2011.per2: 1E-9,
               moisan2011.rper: 1E-10,
               moisan2011.rper2: 1E-9}[per]
        p_exp, s_exp = moisan2011._per(u, inverse_dft=inverse_dft)
        p_act, s_act = per(u, inverse_dft=inverse_dft)
        assert_allclose(p_act, p_exp, eps, eps)
>       assert_allclose(s_act, s_exp, eps, eps)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-09, atol=1e-09
E       
E       Mismatched elements: 1 / 234861 (0.000426%)
E       Max absolute difference: 3.79887751e-06
E       Max relative difference: 5.01928473e-11
E        x: array([[-3.725290e-09+9.396750e-10j, -3.675000e+02+7.568474e+04j,
E               -3.675000e+02+3.784148e+04j, ..., -3.675000e+02-2.522666e+04j,
E               -3.675000e+02-3.784148e+04j, -3.675000e+02-7.568474e+04j],...
E        y: array([[ 0.000000e+00     +0.j      , -3.675000e+02 +75684.742597j,
E               -3.675000e+02 +37841.479069j, ..., -3.675000e+02 -25226.661335j,
E               -3.675000e+02 -37841.479069j, -3.675000e+02 -75684.742597j],...

test_moisan2011.py:134: AssertionError
_____________________________________ test_consistency[per2-u13-False] ______________________________________

per = <function per2 at 0x7fd35cc2d6a8>
u = array([[ 53.,  50.,  52., ...,  69.,  73.,  79.],
       [ 46.,  50.,  53., ...,  96.,  83.,  86.],
       [ 55.,  53.... 76.,  67.,  69.],
       [ 88., 107.,  85., ...,  73.,  69.,  72.],
       [113.,  94., 105., ...,  77.,  81.,  74.]])
inverse_dft = False

    @pytest.mark.parametrize('per, u, inverse_dft', consistency_params())
    def test_consistency(per, u, inverse_dft):
        eps = {moisan2011.per: 1E-11,
               moisan2011.per2: 1E-9,
               moisan2011.rper: 1E-10,
               moisan2011.rper2: 1E-9}[per]
        p_exp, s_exp = moisan2011._per(u, inverse_dft=inverse_dft)
        p_act, s_act = per(u, inverse_dft=inverse_dft)
        assert_allclose(p_act, p_exp, eps, eps)
>       assert_allclose(s_act, s_exp, eps, eps)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-09, atol=1e-09
E       
E       Mismatched elements: 1 / 235508 (0.000425%)
E       Max absolute difference: 3.75950067e-06
E       Max relative difference: 5.24559562e-11
E        x: array([[-3.725290e-09+9.421761e-10j, -3.480000e+02+7.166882e+04j,
E               -3.480000e+02+3.583356e+04j, ..., -3.480000e+02-2.388810e+04j,
E               -3.480000e+02-3.583356e+04j, -3.480000e+02-7.166882e+04j],...
E        y: array([[ 0.000000e+00     +0.j      , -3.480000e+02 +71668.81748j ,
E               -3.480000e+02 +35833.563854j, ..., -3.480000e+02 -23888.103795j,
E               -3.480000e+02 -35833.563854j, -3.480000e+02 -71668.81748j ],...

test_moisan2011.py:134: AssertionError
========================================== short test summary info ==========================================
FAILED test_moisan2011.py::test_mean_s_is_zero[per2-u17-False] - AssertionError: assert 3.841975653468968e...
FAILED test_moisan2011.py::test_mean_s_is_zero[per2-u21-False] - AssertionError: assert 3.842588143336527e...
FAILED test_moisan2011.py::test_mean_s_is_zero[rper2-u33-False] - AssertionError: assert 3.725290298461914...
FAILED test_moisan2011.py::test_mean_s_is_zero[rper2-u37-False] - AssertionError: assert 3.725290298461914...
FAILED test_moisan2011.py::test_consistency[per2-u9-False] - AssertionError: 
FAILED test_moisan2011.py::test_consistency[per2-u13-False] - AssertionError: 
======================================= 6 failed, 138 passed in 3.01s =======================================
```
</details>

sbrisard commented:
I see that these failures are actually tests that some value is zero, and the residual is actually quite small. It is probably okay to play with the tolerances, even if I do not like that too much. Will look into when I have some time.
Thanks again Tobias (and good luck with writing your thesis).
